File: dali/pipeline/operators/arithmetic/arithmetic.cc
// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <vector>

#include "dali/pipeline/operators/arithmetic/arithmetic.h"
#include "dali/kernels/type_tag.h"

namespace dali {

template <>
void ArithmeticGenericOp<CPUBackend>::RunImpl(HostWorkspace &ws) {
  auto &pool = ws.GetThreadPool();
  for (size_t task_idx = 0; task_idx < tile_range_.size(); task_idx++) {
    // TODO(klecki): reduce lambda footprint
    pool.DoWorkWithID([this, &ws, task_idx](int thread_idx) {
      auto range = tile_range_[task_idx];
      // Go over "tiles"
      for (int extent_idx = range.begin; extent_idx < range.end; extent_idx++) {
        // Go over expression tree in some provided order
        for (auto &expr_task : exec_order_) {
          expr_task.impl->Execute(ws, spec_, expr_task.ctx, tile_cover_,
                                  {extent_idx, extent_idx + 1});
        }
      }
    });
  }
}

DALI_SCHEMA(ArithmeticGenericOp)
    .DocStr(R"code(Arithmetic operator capable of executing expression tree of element-wise
 arithmetic operations.)code")
    .AddArg("expression_desc", R"code(Polish notation describing the expression extendend with
 parentheses, see https://en.wikipedia.org/wiki/Polish_notation.
Functions and operations are indicated by names, followed by parentheses.
Inputs (subexpressions) are placed in the parentheses and are separated by spaces,
&<uint> indicates tensor input, $<uint>:<type_string> indicates constant.

More formally using BNF grammar:
<expr>    ::= <call> | <scalar> | <input>
<subexpr> ::= <expr> | <expr> " " <subexpr>
<call>    ::= <name> "(" <subexpr> ")"
<name>    ::= identifier starting with alphabetic character
<input>   ::= "&" <uint>
<scalar>  ::= "$" <uint> ":" <type>
<uint>    ::= unsigned integer
<type>    ::= uint8 | uint16 | uint32 | uint64 | int8 | int16 | int32 | int64
            | float16 | float32 | float64

Examples:
add(&0 mul(&1 $0:int8))
add(&0 rand()))code",
            DALIDataType::DALI_STRING, false)
    .AddOptionalArg("integer_scalars", "", std::vector<int>{})
    .NumInput(1, 64)  // TODO(klecki): Some arbitrary number that needs to be validated in operator
    .AddOptionalArg("float_scalars", "", std::vector<float>{})
    .NumOutput(1)
    .MakeInternal();

DALI_REGISTER_OPERATOR(ArithmeticGenericOp, ArithmeticGenericOp<CPUBackend>, CPU);

}  // namespace dali

File: dali/python/nvidia/dali/experimental/dynamic/_arithmetic.py
# Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import numbers


def _arithm_op(name: str, *args):
    from . import _arithmetic_generic_op
    from ._batch import Batch
    from ._tensor import Tensor, as_tensor

    tensor_args = [arg for arg in args if isinstance(arg, (Tensor, Batch))]
    gpu = any(arg.device.device_type == "gpu" for arg in tensor_args)

    def to_input(arg):
        if not isinstance(arg, (Tensor, Batch)):
            device = "gpu" if gpu and isinstance(arg, (numbers.Real, list, tuple)) else None
            arg = as_tensor(arg, device=device)

        if (arg.device.device_type == "gpu") != gpu:
            raise ValueError("Cannot mix GPU and CPU inputs.")

        return arg

    # only reachable from math functions called with only scalars, e.g. ndd.math.max(2, 3)
    if not tensor_args and args:
        args = (to_input(args[0]), *args[1:])

    desc, inputs, integers, reals = [], [], [], []
    for arg in args:
        type_ = type(arg)
        if type_ is bool:
            desc.append(f"${len(integers)}:bool")
            integers.append(int(arg))
        elif type_ is int:
            if (arg >> 31) not in (0, -1):
                raise OverflowError(f"Integer constant {arg} is out of range for int32.")
            desc.append(f"${len(integers)}:int32")
            integers.append(arg)
        elif type_ is float:
            desc.append(f"${len(reals)}:float32")
            reals.append(arg)
        else:
            desc.append(f"&{len(inputs)}")
            inputs.append(to_input(arg))

    return _arithmetic_generic_op(
        *inputs,
        expression_desc=f"{name}({' '.join(desc)})",
        integer_constants=integers or None,
        real_constants=reals or None,
    )
